src/csexp_rpc/csexp_rpc.ml

```diff
@@ -51,48 +57,72 @@ module Session = struct
     | Some csexp -> Csexp.to_string csexp
 
   let read t =
-    let rec read () =
-      try Csexp.input_opt t.in_channel with
-      | Unix.Unix_error (EBADF, _, _) -> Ok None
-      | Sys_error _ -> Ok None
-      | Sys_blocked_io -> read ()
-      | e -> reraise e
+    let debug res =
+      if debug then Format.eprintf "<< %s@." (string_of_packet res)
     in
-    let+ res = Worker.task t.reader ~f:read in
-    let res =
-      match res with
-      | Error (`Exn exn) ->
-        Worker.stop t.reader;
-        raise exn
-      | Error `Stopped -> None
-      | Ok res -> (
+    match t.reader with
```

Comment:
Do these changes mean that clients will need to wait until they try to read or write to the server before discovering that the server has died/closed the connection?

Reply:
Yes, I don't see how else it can be possible. Note that it's not a problem in practice since the client is always blocking on a read so the disconnect should be come promptly.